### PR TITLE
unionized material energy-cross section grid lookup

### PR DESCRIPTION
**Pull request by walshjon**
This PR introduces an energy-cross section grid lookup capability that uses energy-unionized grids by material.  This has many similarities to the unionized grid algorithm, previously removed from OpenMC, which unionized grids across all nuclides in the model.  That algorithm was closely related unionization capabilites previously implemented in other codes (e.g. RACER, Serpent).

This variant is introduced because there are advantages to unionizing on a material-by-material basis.  Considering a depleted LWR pin cell, when a neutron is in the moderator, undergoing events and grid lookups, we don't want to have to search over a grid that includes all of the energies from the grids of the nuclides in the depleted fuel.

Serial speedups of ~1.25 and ~1.5 relative to the default logarithmic hash search and individual nuclide grid search, respectively, are observed for an LWR pin cell problem with >150 nuclides, ENDF71x data.  Speedup in multi-threaded runs is not appreciably degraded.

### CODE REVIEW

File: src/input_xml.F90 [hunk 1]
```diff
@@ -213,8 +213,8 @@ subroutine read_settings_xml()
     select case (trim(temp_str))
     case ('nuclide')
       grid_method = GRID_NUCLIDE
-    case ('union')
-      call fatal_error("Union energy grid is no longer supported.")
```
Comment (paulromano) [inline src/input_xml.F90]: Maybe we could have 'union' default to the material union grid with a warning message that the traditional union energy grid method no longer exists?

File: src/cross_section.F90 [hunk 2]
```diff
@@ -48,6 +51,10 @@ subroutine calculate_xs(p)
 
     mat => materials(p % material)
 
+    ! Find energy index on global or material unionized grid
+    if (grid_method == GRID_MAT_UNION) &
+      & call find_energy_index(p % E, p % material)
```
Comment (paulromano) [inline src/cross_section.F90]: The ampersand on the second line is superfluous (only needed if you are continuing a string). The style guide says continuation lines are indented with 5 extra spaces.

File: src/initialize.F90 [hunk 3]
```diff
@@ -109,10 +109,19 @@ subroutine initialize_run()
...
-      if (grid_method == GRID_LOGARITHM) then
+      ! Construct unionized or log energy grid for cross-sections
+      select case (grid_method)
+      case (GRID_NUCLIDE)
+        continue
+      case (GRID_MAT_UNION)
+        call time_unionize % start()
+!$omp critical
+        call unionized_grid()
+!$omp end critical
```
Comment (paulromano) [inline src/initialize.F90]: This is not within an `!$omp parallel` region, so I don't see why a critical section is needed.

**Comment by paulromano** [comment]
You'll need to add a test in the test suite which uses the material-union method.

**Comment by walshjon** [comment]
Thanks for the review @paulromano.  I believe I've addressed your suggestions with the subsequent commits.

**Comment by paulromano** [comment]
Sorry for the delay -- I'm back in business now. Two last things: update the documentation to include "material-union" as an option for `<energy_grid>`, and update the RELAX-NG schemata for settings.xml (edit the .rnc file and then generate the .rng by running `trang settings.rnc settings.rng`).

**Comment by paulromano** [comment]
Thanks @walshjon!

**Pull request merged**